[WebCore] Add big endian bitmap texture

sideeffect42 opened the pull request:
#### c37138cb8d45196e3379961b1f0a91fc30541b2d
<pre>
[WebCore] Add big endian bitmap texture
<a href="https://bugs.webkit.org/show_bug.cgi?id=245446">https://bugs.webkit.org/show_bug.cgi?id=245446</a>

Reviewed by Michael Catanzaro and Miguel Gomez.

Adding the ARGB bitmap for big endian fixes colours when WEBKIT_DISABLE_COMPOSITING_MODE=1 is set.

* Source/WebCore/platform/graphics/texmap/BitmapTexture.cpp: add bitmap texture for big endian.

Canonical link: <a href="https://commits.webkit.org/275780@main">https://commits.webkit.org/275780@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/43c044332d653f8f6bc9a89664d6afbd564acab4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42525 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21544 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44922 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45129 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38645 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/44832 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/24785 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18905 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35224 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43099 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18464 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36604 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16150 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/37660 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/581 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/38702 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37986 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46610 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17343 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Sonoma-Debug-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/41893 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18961 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40505 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9552 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19025 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18606 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebCore/platform/graphics/texmap/BitmapTexture.cpp:
```diff
@@ -86,7 +86,11 @@ void BitmapTexture::updateContents(const void* srcData, const IntRect& targetRec
     // We are updating a texture with format RGBA with content from a buffer that has BGRA format. Instead of turning BGRA
     // into RGBA and then uploading it, we upload it as is. This causes the texture format to be RGBA but the content to be BGRA,
     // so we mark the texture to convert the colors when painting the texture.
+#if G_BYTE_ORDER == G_LITTLE_ENDIAN
```
Looks like this doesn't build on Windows. This file is cross-platform and shouldn't be using GLib stuff.

sideeffect42 replied:
Yes, I just saw this, too.

sideeffect42 replied:
Let's see if Windows has `__BYTE_ORDER__`…

mcatanzaro replied:
Seems that worked.

Constellation commented on Source/WebCore/platform/graphics/texmap/BitmapTexture.cpp:
```diff
@@ -86,7 +86,11 @@ void BitmapTexture::updateContents(const void* srcData, const IntRect& targetRec
     // We are updating a texture with format RGBA with content from a buffer that has BGRA format. Instead of turning BGRA
     // into RGBA and then uploading it, we upload it as is. This causes the texture format to be RGBA but the content to be BGRA,
     // so we mark the texture to convert the colors when painting the texture.
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
```
Use `CPU(LITTLE_ENDIAN)`

sideeffect42 replied:
I changed it.

Would you mind to explain why `CPU(LITTLE_ENDIAN)` is preferred over the various other  ways to handle endianness-conditional code across the WebKit codebase?

mcatanzaro replied:
It's an abstraction that can be implemented in terms of anything else, dependent on platform. Maybe it uses `__BYTE_ORDER__` or maybe it uses `G_BYTE_ORDER`. Doesn't matter.

Commit:
[WebCore] Add big endian bitmap texture
https://bugs.webkit.org/show_bug.cgi?id=245446

Reviewed by Michael Catanzaro and Miguel Gomez.

Adding the ARGB bitmap for big endian fixes colours when WEBKIT_DISABLE_COMPOSITING_MODE=1 is set.

* Source/WebCore/platform/graphics/texmap/BitmapTexture.cpp: add bitmap texture for big endian.

Canonical link: https://commits.webkit.org/275780@main
